File: src/oauthkitchen/analyzers/translator.py
"""Permission Translator - converts raw permissions to plain English with risk context."""

from __future__ import annotations

from dataclasses import dataclass
from pathlib import Path
from typing import Any

import yaml

from oauthkitchen.models import PermissionDefinition, PermissionType, RiskCategory
from oauthkitchen.utils.logging import get_logger

logger = get_logger("translator")

# Default rules file path
DEFAULT_RULES_PATH = Path(__file__).parent.parent / "rules" / "permissions.yaml"


@dataclass
class TranslatedPermission:
    """Result of translating a permission."""
    permission: str
    resource: str
    plain_english: str
    category: RiskCategory
    category_label: str
    impact_score: int
    abuse_scenarios: list[str]
    admin_impact_note: str | None
    is_known: bool


class PermissionTranslator:
    """
    Translates OAuth permission names into plain English with risk context.

    Uses a community-editable rules file (YAML) that maps permission names
    to descriptions, categories, and abuse scenarios.
    """

    # Category display labels
    CATEGORY_LABELS = {
        RiskCategory.READ_ONLY: "Read-only",
        RiskCategory.DATA_EXFILTRATION: "Data exfiltration",
        RiskCategory.PRIVILEGE_ESCALATION: "Privilege escalation",
        RiskCategory.TENANT_TAKEOVER: "Tenant takeover potential",
        RiskCategory.PERSISTENCE: "Persistence",
        RiskCategory.LATERAL_MOVEMENT: "Lateral movement",
        RiskCategory.UNKNOWN: "Unknown",
    }

    def __init__(self, rules_path: Path | str | None = None):
        """
        Initialize the translator.

        Args:
            rules_path: Path to the rules YAML file. Uses default if not specified.
        """
        self.rules_path = Path(rules_path) if rules_path else DEFAULT_RULES_PATH
        self.rules: dict[str, dict[str, Any]] = {}
        self._load_rules()

    def _load_rules(self) -> None:
        """Load rules from the YAML file."""
        if not self.rules_path.exists():
            logger.warning(
                "Rules file not found at %s, using empty rules",
                self.rules_path
            )
            return

        try:
            with open(self.rules_path, "r", encoding="utf-8") as f:
                data = yaml.safe_load(f) or {}

            # Flatten the resource-grouped structure
            for resource, permissions in data.items():
                if isinstance(permissions, dict):
                    for perm_name, perm_data in permissions.items():
                        # Store with resource prefix for lookup
                        key = perm_name.lower()
                        self.rules[key] = {
                            "resource": resource,
                            **perm_data
                        }

            logger.info("Loaded %d permission rules", len(self.rules))

        except (yaml.YAMLError, OSError) as e:
            logger.error("Failed to load rules file: %s", e)

    def translate(
        self,
        permission: str,
        resource: str = "microsoft_graph"
    ) -> TranslatedPermission:
        """
        Translate a single permission.

        Args:
            permission: The permission name (e.g., "Mail.Read")
            resource: The resource API (e.g., "microsoft_graph")

        Returns:
            TranslatedPermission with all context
        """
        # Normalize permission name for lookup
        key = permission.lower()

        rule = self.rules.get(key)

        if rule:
            category = self._parse_category(rule.get("category", "unknown"))
            return TranslatedPermission(
                permission=permission,
                resource=rule.get("resource", resource),
                plain_english=rule.get(
                    "plain_english",
                    rule.get("display_name", permission)
                ),
                category=category,
                category_label=self.CATEGORY_LABELS.get(category, "Unknown"),
                impact_score=rule.get("impact_score", 30),
                abuse_scenarios=rule.get("abuse_scenarios", []),
                admin_impact_note=rule.get("admin_impact_note"),
                is_known=True,
            )

        # Unknown permission - return with sensible defaults
        return TranslatedPermission(
            permission=permission,
            resource=resource,
            plain_english=f"Permission: {permission} (no translation available)",
            category=RiskCategory.UNKNOWN,
            category_label="Unknown - requires review",
            impact_score=30,  # Moderate default
            abuse_scenarios=[],
            admin_impact_note=None,
            is_known=False,
        )

    def translate_many(
        self,
        permissions: list[str],
        resource: str = "microsoft_graph"
    ) -> list[TranslatedPermission]:
        """Translate multiple permissions."""
        return [self.translate(p, resource) for p in permissions]

    def enrich_permission_definition(
        self,
        permission: PermissionDefinition
    ) -> PermissionDefinition:
        """
        Enrich a PermissionDefinition with translation data.

        Modifies the permission in place and returns it.
        """
        translated = self.translate(permission.value)

        permission.plain_english = translated.plain_english
        permission.risk_category = translated.category
        permission.abuse_scenarios = translated.abuse_scenarios
        permission.admin_impact_note = translated.admin_impact_note
        permission.impact_score = translated.impact_score

        return permission

    def get_high_impact_permissions(
        self,
        min_score: int = 70
    ) -> list[tuple[str, TranslatedPermission]]:
        """
        Get all known high-impact permissions.

        Args:
            min_score: Minimum impact score to include

        Returns:
            List of (permission_name, TranslatedPermission) tuples
        """
        results = []
        for perm_key, rule in self.rules.items():
            if rule.get("impact_score", 0) >= min_score:
                # Reconstruct the original case from display_name or use key
                perm_name = rule.get("display_name", perm_key)
                # Find original casing
                for orig_key in self.rules:
                    if orig_key.lower() == perm_key:
                        perm_name = orig_key
                        break

                translated = self.translate(perm_name)
                results.append((perm_name, translated))

        return sorted(results, key=lambda x: x[1].impact_score, reverse=True)

    def format_permission_report(
        self,
        permission: str,
        include_scenarios: bool = True
    ) -> str:
        """
        Format a permission into a human-readable report string.

        Args:
            permission: Permission to translate
            include_scenarios: Whether to include abuse scenarios
        """
        translated = self.translate(permission)

        lines = [
            f"Permission: {translated.permission}",
            f"Resource: {translated.resource}",
            f"Description: {translated.plain_english}",
            f"Risk Category: {translated.category_label}",
            f"Impact Score: {translated.impact_score}/100",
        ]

        if translated.admin_impact_note:
            lines.append(f"Admin Impact: {translated.admin_impact_note}")

        if include_scenarios and translated.abuse_scenarios:
            lines.append("Potential Abuse Scenarios:")
            for scenario in translated.abuse_scenarios:
                lines.append(f"  • {scenario}")

        if not translated.is_known:
            lines.append("")
            lines.append("⚠ This permission is not in the rules database.")
            lines.append("  Please review the raw permission documentation.")

        return "\n".join(lines)

    def _parse_category(self, category_str: str) -> RiskCategory:
        """Parse category string to RiskCategory enum."""
        mapping = {
            "read_only": RiskCategory.READ_ONLY,
            "data_exfiltration": RiskCategory.DATA_EXFILTRATION,
            "privilege_escalation": RiskCategory.PRIVILEGE_ESCALATION,
            "tenant_takeover": RiskCategory.TENANT_TAKEOVER,
            "persistence": RiskCategory.PERSISTENCE,
            "lateral_movement": RiskCategory.LATERAL_MOVEMENT,
        }
        return mapping.get(category_str.lower(), RiskCategory.UNKNOWN)

    @property
    def known_permission_count(self) -> int:
        """Number of permissions in the rules database."""
        return len(self.rules)

    def reload_rules(self, rules_path: Path | str | None = None) -> None:
        """Reload rules from file."""
        if rules_path:
            self.rules_path = Path(rules_path)
        self.rules.clear()
        self._load_rules()

File: src/lib/analyzers/translator.ts
/**
 * Permission Translator - converts raw permissions to plain English with risk context.
 */

import { RiskCategory } from '@/types/models'
import { getLogger } from '@/lib/utils/logger'

const logger = getLogger('translator')

// Default rules file path
const DEFAULT_RULES_PATH = '/public/permissions.json'

export interface TranslatedPermission {
  permission: string
  resource: string
  plainEnglish: string
  category: RiskCategory
  categoryLabel: string
  impactScore: number
  abuseScenarios: string[]
  adminImpactNote: string | null
  isKnown: boolean
}

export interface PermissionRule {
  displayName?: string
  plainEnglish?: string
  category?: string
  impactScore?: number
  abuseScenarios?: string[]
  adminImpactNote?: string
  resource?: string
  [key: string]: unknown
}

/**
 * Translates OAuth permission names into plain English with risk context.
 *
 * Uses a rules file (JSON) that maps permission names to descriptions,
 * categories, and abuse scenarios.
 */
export class PermissionTranslator {
  private rulesPath: string
  private rules: Map<string, PermissionRule> = new Map()
  private isLoaded = false
  private loadPromise: Promise<void> | null = null

  private static readonly CATEGORY_LABELS: Record<RiskCategory, string> = {
    [RiskCategory.READ_ONLY]: 'Read-only',
    [RiskCategory.DATA_EXFILTRATION]: 'Data exfiltration',
    [RiskCategory.PRIVILEGE_ESCALATION]: 'Privilege escalation',
    [RiskCategory.TENANT_TAKEOVER]: 'Tenant takeover potential',
    [RiskCategory.PERSISTENCE]: 'Persistence',
    [RiskCategory.LATERAL_MOVEMENT]: 'Lateral movement',
    [RiskCategory.UNKNOWN]: 'Unknown',
  }

  constructor(rulesPath: string = DEFAULT_RULES_PATH) {
    this.rulesPath = rulesPath
  }

  /**
   * Load rules from the JSON file.
   */
  async loadRules(): Promise<void> {
    if (this.isLoaded) return
    if (this.loadPromise) return this.loadPromise

    this.loadPromise = this._loadRulesInternal()
    await this.loadPromise
  }

  private async _loadRulesInternal(): Promise<void> {
    try {
      const response = await fetch(this.rulesPath)
      if (!response.ok) {
        logger.warn(`Rules file not found at ${this.rulesPath}, using empty rules`)
        this.isLoaded = true
        return
      }

      const data = await response.json()

      // Flatten the resource-grouped structure
      for (const resource in data) {
        if (typeof data[resource] === 'object' && data[resource] !== null) {
          const permissions = data[resource] as Record<string, PermissionRule>
          for (const permName in permissions) {
            const permData = permissions[permName]
            const key = permName.toLowerCase()
            this.rules.set(key, {
              resource,
              ...permData,
            })
          }
        }
      }

      logger.info(`Loaded ${this.rules.size} permission rules`)
      this.isLoaded = true
    } catch (error) {
      logger.error(
        `Failed to load rules file: ${error instanceof Error ? error.message : String(error)}`
      )
      this.isLoaded = true
    }
  }

  /**
   * Translate a single permission.
   */
  translate(permission: string, resource: string = 'microsoft_graph'): TranslatedPermission {
    // Normalize permission name for lookup
    const key = permission.toLowerCase()
    const rule = this.rules.get(key)

    if (rule) {
      const category = this._parseCategory(rule.category || 'unknown')
      return {
        permission,
        resource: rule.resource || resource,
        plainEnglish: rule.plainEnglish || rule.displayName || permission,
        category,
        categoryLabel: PermissionTranslator.CATEGORY_LABELS[category],
        impactScore: rule.impactScore || 30,
        abuseScenarios: rule.abuseScenarios || [],
        adminImpactNote: rule.adminImpactNote || null,
        isKnown: true,
      }
    }

    // Unknown permission - return with sensible defaults
    return {
      permission,
      resource,
      plainEnglish: `Permission: ${permission} (no translation available)`,
      category: RiskCategory.UNKNOWN,
      categoryLabel: 'Unknown - requires review',
      impactScore: 30,
      abuseScenarios: [],
      adminImpactNote: null,
      isKnown: false,
    }
  }

  /**
   * Translate multiple permissions.
   */
  translateMany(
    permissions: string[],
    resource: string = 'microsoft_graph'
  ): TranslatedPermission[] {
    return permissions.map((p) => this.translate(p, resource))
  }

  /**
   * Get all known high-impact permissions.
   */
  getHighImpactPermissions(minScore: number = 70): Array<[string, TranslatedPermission]> {
    const results: Array<[string, TranslatedPermission]> = []

    for (const [permKey, rule] of this.rules.entries()) {
      if ((rule.impactScore || 0) >= minScore) {
        const permName = rule.displayName || permKey
        const translated = this.translate(permName)
        results.push([permName, translated])
      }
    }

    return results.sort((a, b) => b[1].impactScore - a[1].impactScore)
  }

  /**
   * Format a permission into a human-readable report string.
   */
  formatPermissionReport(permission: string, includeScenarios: boolean = true): string {
    const translated = this.translate(permission)

    const lines: string[] = [
      `Permission: ${translated.permission}`,
      `Resource: ${translated.resource}`,
      `Description: ${translated.plainEnglish}`,
      `Risk Category: ${translated.categoryLabel}`,
      `Impact Score: ${translated.impactScore}/100`,
    ]

    if (translated.adminImpactNote) {
      lines.push(`Admin Impact: ${translated.adminImpactNote}`)
    }

    if (includeScenarios && translated.abuseScenarios.length > 0) {
      lines.push('Potential Abuse Scenarios:')
      for (const scenario of translated.abuseScenarios) {
        lines.push(`  • ${scenario}`)
      }
    }

    if (!translated.isKnown) {
      lines.push('')
      lines.push('⚠ This permission is not in the rules database.')
      lines.push('  Please review the raw permission documentation.')
    }

    return lines.join('\n')
  }

  /**
   * Parse category string to RiskCategory enum.
   */
  private _parseCategory(categoryStr: string): RiskCategory {
    const mapping: Record<string, RiskCategory> = {
      read_only: RiskCategory.READ_ONLY,
      data_exfiltration: RiskCategory.DATA_EXFILTRATION,
      privilege_escalation: RiskCategory.PRIVILEGE_ESCALATION,
      tenant_takeover: RiskCategory.TENANT_TAKEOVER,
      persistence: RiskCategory.PERSISTENCE,
      lateral_movement: RiskCategory.LATERAL_MOVEMENT,
    }
    return mapping[categoryStr.toLowerCase()] || RiskCategory.UNKNOWN
  }

  /**
   * Get the number of permissions in the rules database.
   */
  getKnownPermissionCount(): number {
    return this.rules.size
  }

  /**
   * Reload rules from file.
   */
  async reloadRules(rulesPath?: string): Promise<void> {
    if (rulesPath) {
      this.rulesPath = rulesPath
    }
    this.rules.clear()
    this.isLoaded = false
    this.loadPromise = null
    await this.loadRules()
  }
}
